numba/np/ufunc/gufunc_scheduler.cpp

```diff
@@ -68,8 +78,35 @@ class RangeActual {
         }
         return ret;
     }
+
+    uintp total_size() const {
+        std::vector<intp> per_dim = iters_per_dim();
+        uintp res = 1;
+        for (unsigned i = 0; i < per_dim.size(); ++i) {
+            res *= per_dim[i];
+        }
+        return res;
+    }
 };
 
+extern "C" void set_parallel_chunksize(uintp n) {
+    parallel_chunksize = n;
+}
+
+extern "C" uintp get_parallel_chunksize() {
+    return parallel_chunksize;
+}
+
+extern "C" uintp get_sched_size(uintp num_threads, uintp num_dim, intp *starts, intp *ends) {
+    if (parallel_chunksize == 0) {
+        return num_threads;
+    }
+    RangeActual ra(num_dim, starts, ends);
+    uintp total_work_size = ra.total_size();
+    uintp num_divisions = total_work_size / parallel_chunksize;
+    return num_divisions < num_threads ? num_threads : num_divisions;
```

Comment:
@stuartarchibald We already deviate from the OpenMP model because we have multi-dimensional parfors (and the chunksize has to apply to apply to the combination of those dimensions).  I'm not clear on this OpenMP point of how they could have nested parfors with different kinds of schedules potentially in combination with the collapse option.  Anyway,  I was inspired by OpenMP but not bound to it.  The chunksize you request is actually approximate.  As in the above code, we first turn the chunksize into the number of divisions of the work.  Do you agree with line 107?  If the array is 100 long and they have 4 threads and they specify a chunksize of 100 then do we use 4 threads or just 1?

If you had a 2D parfor and specified a chunksize of 5 then you are never going to get any chunk of work with exactly 5  items in it.  The system will try to make the average 5 but some will be larger and some smaller (4 or 6).

For your example, our system would currently see that 1 < 10/7 < 2 and so would need two divisions of the work. Then, it would do 2 equal divisions so you'd get 5 and 5.

Reply:
I think this makes sense as it's bound to parfors, particularly RE 2d parfor. Also agree with l107, but think this example should go into the docs as I think it's quite likely a user would encounter it?

Reply:
Done.